### PR TITLE
Tradheli: New integrator management implementation

### PR DESCRIPTION
**Pull request by bnsgeyer**
This PR provides a selectable option for a new integrator management scheme for the Pitch, Roll, and Yaw axes.  The existing scheme is retained as default.  The new scheme uses collective and the landing detector as the triggers for releasing and resetting the integrator.  As this relies heavily on the landing detector, this was provided as an option initially to allow users to test the feature and disable it if they were seeing issues with the landing detector.

What has changed:
-New integrator scheme no longer uses leaky I implementation or minimum leak (ILMI) for the integrator when the vehicle speed is less than 5 m/s.
-IMAX is only integrator limit but growth is managed for takeoff/landing.
-Hover collective learning was added based off of multicopter hover throttle learning
-Takeoff
        - integrator is reset until liftoff is detected
        - lift off is detected using collective output (in motors class).  Must be greater than 50% of hover collective
        - hover collective is either manually specified through new parameter or determined by hover collective learning

-Landing
        - when landing the integrator is frozen when land_complete_maybe flag is set
        - integrator is smoothly reset to zero over 0.5 seconds once landing_complete flag is set
        - integrator remains reset until takeoff condition is met above

The new integrator scheme was tested both in the sim and on an actual helicopter.  My recent addition of making it selectable still needs a little more testing to ensure nothing was broken.

### CODE REVIEW

File: libraries/AP_Motors/AP_MotorsHeli.h [hunk 1]
```diff
@@ -203,6 +227,8 @@ class AP_MotorsHeli : public AP_Motors {
         uint8_t in_autorotation         : 1;    // true if aircraft is in autorotation
         uint8_t enable_bailout          : 1;    // true if allowing RSC to quickly ramp up engine
         uint8_t servo_test_running      : 1;    // true if servo_test is running
+        uint8_t land_complete           : 1;    // true if aircraft is landed
```
Comment (tridge) [inline libraries/AP_Motors/AP_MotorsHeli.h:230]: this makes heliflags 16 bits, need to check that is OK (is it logged somewhere as 8 bit?)

File: libraries/AP_Motors/AP_MotorsHeli.cpp [hunk 2]
```diff
@@ -96,6 +96,27 @@ const AP_Param::GroupInfo AP_MotorsHeli::var_info[] = {
...
+    // @Values: 0:Disabled, 1:Learn, 2:Learn and Save
+    // @User: Advanced
+    AP_GROUPINFO("HOVER_LEARN", 27, AP_MotorsHeli, _collective_hover_learn, HOVER_LEARN_AND_SAVE),
+
+    // @Param: I_TERM_MGMT
+    // @DisplayName: Integrator Management
+    // @Description: Allows user to change how the pitch, roll, and yaw axis integrator term is managed for low speed and takeoff/landing. The original scheme uses a leaky integrator for ground speeds less than 5 m/s and won't let the steady state integrator build above ILMI. The integrator is allowed to build to the ILMI value when it is landed.  The new scheme bases integrator limiting on takeoff and landing.  Whenever the aircraft is landed the integrator is set to zero.  When the aicraft is airborne, the integrator is only limited by IMAX.  It also uses hover collective learning to determine when to release the integrator for takeoff. 
+    // @Values: 0:Original, 1:New
+    // @User: Standard
+    AP_GROUPINFO("I_TERM_MGMT", 28, AP_MotorsHeli, _new_integrator, 0),
```
Comment (tridge) [inline libraries/AP_Motors/AP_MotorsHeli.cpp:118]: maybe make a HELI_OPTIONS param as a bitmask as a AP_Int32

**Review by tridge — CHANGES_REQUESTED** [review]

File: libraries/AP_Motors/AP_MotorsHeli.h [hunk 3]
```diff
@@ -119,7 +123,11 @@ class AP_MotorsHeli : public AP_Motors {
     // supports_yaw_passthrough
     virtual bool supports_yaw_passthrough() const { return false; }
 
-    float get_throttle_hover() const override { return 0.5f; }
+    // update estimated throttle required to hover
+    void update_throttle_hover(float dt);
+    float get_throttle_hover() const override { return _collective_hover; }
+
+    bool get_takeoff_collective() const { return _heliflags.takeoff_collective; }
```
Comment (rmackay9) [inline libraries/AP_Motors/AP_MotorsHeli.h:130]: maybe put a comment get_takeoff_colective()

File: libraries/AP_Motors/AP_MotorsHeli.h [hunk 4]
```diff
@@ -212,6 +238,9 @@ class AP_MotorsHeli : public AP_Motors {
     AP_Int16        _collective_mid;            // Swash servo position corresponding to zero collective pitch (or zero lift for Asymmetrical blades)
     AP_Int8         _servo_mode;                // Pass radio inputs directly to servos during set-up through mission planner
     AP_Int8         _servo_test;                // sets number of cycles to test servo movement on bootup
+    AP_Float        _collective_hover;          // estimated collective required to hover throttle in the range 0 ~ 1
+    AP_Int8         _collective_hover_learn;    // enable/disabled hover collective learning
+    AP_Int8         _new_integrator;     // enable/disabled new integrator management scheme
```
Comment (rmackay9) [inline libraries/AP_Motors/AP_MotorsHeli.h:243]: As discussed on the call, let's not use the term, "new".  So either "leaky" or come up with some other name.

File: ArduCopter/mode_acro_heli.cpp [hunk 5]
```diff
@@ -49,16 +49,16 @@ void ModeAcro_Heli::run()
         attitude_control->reset_rate_controller_I_terms();
         break;
     case AP_Motors::SpoolState::GROUND_IDLE:
-        // Landed
-        if (motors->init_targets_on_arming()) {
+        // If aircraft is landed, set target heading to current and reset the integrator
+        // Otherwise motors could be at ground idle for practice autorotation
+        if ((motors->init_targets_on_arming() && !motors->use_new_integrator_mgmt()) || (copter.ap.land_complete && motors->use_new_integrator_mgmt())) {
```
Comment (bnsgeyer) [inline ArduCopter/mode_acro_heli.cpp:54]: modify parameter name.  Maybe !use_leaky_integrator()

Comment (rmackay9) [reply ArduCopter/mode_acro_heli.cpp:54]: yes, that sounds good to me

File: libraries/AC_PID/AC_PID.cpp [hunk 6]
```diff
@@ -261,7 +261,19 @@ float AC_PID::get_ff()
 
 void AC_PID::reset_I()
 {
-    _integrator = 0;
```
Comment (Gone4Dirt) [inline libraries/AC_PID/AC_PID.cpp:264]: Isn't this going to change the behaviour of all AC_PID objects that use reset I?  Is that the intention?

File: libraries/AP_Motors/AP_MotorsHeli.h (hunk 1 above)
Comment (bnsgeyer) [reply libraries/AP_Motors/AP_MotorsHeli.h:238]: From what I can tell, only accessors to individual flags are used.

File: libraries/AC_PID/AC_PID.cpp (hunk 6 above)
Comment (bnsgeyer) [reply libraries/AC_PID/AC_PID.cpp:264]: Yes.  From discussion with @lthall, he thought this was acceptable.  Most of the time that multi's are using the reset, they aren't spinning props so this shouldn't affect them.  It is really noticeable in heli's.  You may have already noticed this in Loiter with the current Leaky I and ILMI implementation.

File: libraries/AC_PID/AC_PID.cpp [hunk 7]
```diff
@@ -261,7 +261,19 @@ float AC_PID::get_ff()
...
+    float reset_time = AC_PID_RESET_TC * 3.0f;
+    uint64_t now = AP_HAL::micros64();
+
+    if ((float)(now - _reset_last_update)*0.000001 > 0.50f ) {
```
[the remark this answers is not in the document]
Comment (bnsgeyer) [reply libraries/AC_PID/AC_PID.cpp:267]: done. Thanks!

File: libraries/AP_Motors/AP_MotorsHeli.cpp [hunk 8]
```diff
@@ -301,6 +323,17 @@ void AP_MotorsHeli::output_logic()
             // Motors should be stationary.
             // Servos set to their trim values or in a test condition.
 
+            // set limits flags
+            if (!using_leaky_integrator()) {
+                limit.roll = true;
```
[the remark this answers is not in the document]
Comment (bnsgeyer) [reply libraries/AP_Motors/AP_MotorsHeli.cpp:328]: done.  Thanks!

File: libraries/AP_Motors/AP_MotorsHeli_Dual.cpp [hunk 9]
```diff
@@ -560,6 +557,12 @@ void AP_MotorsHeli_Dual::move_actuators(float roll_out, float pitch_out, float c
         limit.throttle_lower = true;
     }
 
+    if (collective_out > _collective_mid_pct + 0.5f * (_collective_hover - _collective_mid_pct)) {
```
[the remark this answers is not in the document]
Comment (bnsgeyer) [reply libraries/AP_Motors/AP_MotorsHeli_Dual.cpp]: done.  Thanks!

File: libraries/AP_Motors/AP_MotorsHeli.cpp [hunk 10]
```diff
@@ -456,3 +533,34 @@ void AP_MotorsHeli::rc_write_swash(uint8_t chan, float swash_in)
...
+{
+    if (_collective_hover_learn != HOVER_LEARN_DISABLED) {
+
+        // Don't let _collective_hover go below H_COLL_MID
+        float curr_collective = get_throttle();
+        if (curr_collective < _collective_mid_pct) {
+            curr_collective = _collective_mid_pct;
+        }
+
+        // we have chosen to constrain the hover throttle to be within the range reachable by the third order expo polynomial.
```
[the remark this answers is not in the document]
Comment (bnsgeyer) [reply libraries/AP_Motors/AP_MotorsHeli.cpp]: Done.

File: libraries/AP_Motors/AP_MotorsHeli.h (hunk 3 above)
Comment (bnsgeyer) [reply libraries/AP_Motors/AP_MotorsHeli.h:130]: done.  thanks!

File: libraries/AP_Motors/AP_MotorsHeli.cpp [hunk 11]
```diff
@@ -96,6 +96,28 @@ const AP_Param::GroupInfo AP_MotorsHeli::var_info[] = {
...
+    // @User: Advanced
+    AP_GROUPINFO("HOVER_LEARN", 27, AP_MotorsHeli, _collective_hover_learn, HOVER_LEARN_AND_SAVE),
+
+    // @Param: OPTIONS
+    // @DisplayName: Heli_Options
+    // @Description: Bitmask of heli options.  Bit 0 changes how the pitch, roll, and yaw axis integrator term is managed for low speed and takeoff/landing. In AC 4.0 and earlier, scheme uses a leaky integrator for ground speeds less than 5 m/s and won't let the steady state integrator build above ILMI. The integrator is allowed to build to the ILMI value when it is landed.  The other integrator management scheme bases integrator limiting on takeoff and landing.  Whenever the aircraft is landed the integrator is set to zero.  When the aicraft is airborne, the integrator is only limited by IMAX. 
+    // @Values: 0:I term management based landed state, 1:Leaky I(4.0 and earlier)
+    // @Bitmask: 0:Use Leaky I
+    // @User: Standard
+    AP_GROUPINFO("OPTIONS", 28, AP_MotorsHeli, _heli_options, (float)HeliOption::USE_LEAKY_I),
```
[the remark this answers is not in the document]
Comment (bnsgeyer) [reply libraries/AP_Motors/AP_MotorsHeli.cpp]: done.  thanks!

File: libraries/AC_PID/AC_PID.cpp (hunk 6 above)
Comment (lthall) [reply libraries/AC_PID/AC_PID.cpp:264]: Yep, happy with this.

**Review by lthall — APPROVED** [review]
Good to go from my end!

**Comment by bnsgeyer** [comment]
@rmackay9 Per our conversation a few hours ago, I incorporated changes to the land detector as discussed.  I decided to move forward with the new PID function for smoothly resetting the I term.  Included a helper function in Attitude Control library and then used it in Copter where I felt it was necessary.  I know that this is different from what we discussed but after looking at it, I decided it was worth trying to get it in with this PR.  As you can see in the approval, @lthall already approved of the concept.  Now I just broke it out so we are more confident that it isn't being used somewhere it shouldn't.

**Review by tridge — APPROVED** [review]

**Review by rmackay9 — APPROVED** [review]

**Pull request merged**